libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+        let mask = (BigUint::from(3u8) << (bits - 2)) | BigUint::one();
+
+        loop {
+            res = rng.gen_biguint_range(&start, &end);
+            res |= &mask;
+
+            if prime::check(&res) {
+                debug!("Found prime in {} iteration", iteration);
+                break;
+            }
+            iteration += 1;
+        }
+
+        match res.to_bigint() {
+            Some(bn) => Ok(BigNumber{bn}),
+            None => Err(UrsaCryptoError::InvalidStructure("Unable to generate prime in range".to_string()))
+        }
+    }
+
+    pub fn is_prime(&self, _ctx: Option<&mut BigNumberContext>) -> Result<bool, UrsaCryptoError> {
+        if self.is_negative() {
+            Ok(false)
+        } else {
+            match self.bn.to_biguint() {
+                Some(bn) => Ok(prime::check(&bn)),
+                None => Err(UrsaCryptoError::InvalidStructure("An error in is_prime".to_string()))
+            }
+        }
+    }
+
+    pub fn is_safe_prime(&self, _ctx: Option<&mut BigNumberContext>) -> Result<bool, UrsaCryptoError> {
+        if self.bn < BigInt::zero() {
+            Ok(false)
+        } else {
+            match self.bn.to_biguint() {
+                Some(bn) => Ok(safe_prime::check(&bn)),
+                None => Err(UrsaCryptoError::InvalidStructure("An error in is_safe_prime".to_string()))
+            }
+        }
+    }
+
+    pub fn rand(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        let mut rng = OsRng::new()?;
+        let res = rng.gen_biguint(size).to_bigint();
+        Ok(BigNumber { bn: res.unwrap() })
+    }
+
+    pub fn rand_range(&self) -> Result<BigNumber, UrsaCryptoError> {
+        let mut rng = OsRng::new()?;
+        let res = rng.gen_bigint_range(&BigInt::zero(), &self.bn);
+        match res.to_bigint() {
+            Some(bn) => Ok(BigNumber{bn}),
+            None => Err(UrsaCryptoError::InvalidStructure("An error in rand_range".to_string()))
+        }
+    }
+
+    pub fn num_bits(&self) -> Result<i32, UrsaCryptoError> {
+        Ok(self.bn.bits() as i32)
+    }
+
+    pub fn is_bit_set(&self, n: i32) -> Result<bool, UrsaCryptoError> {
+        let bits = n as usize;
+        let res = &self.bn >> bits;
+        Ok(res.is_odd())
+    }
+
+    pub fn set_bit(&mut self, n: i32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        let bits = n as usize;
+        let mask = BigInt::one() << bits;
+        self.bn |= mask;
+        Ok(self)
+    }
+
+    pub fn from_u32(n: usize) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: BigInt::from(n) })
+    }
+
+    pub fn from_dec(dec: &str) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber{ bn: BigInt::from_str_radix(dec, 10)? })
+    }
+
+    pub fn from_hex(hex: &str) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber{ bn: BigInt::from_str_radix(hex, 16)? })
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<BigNumber, UrsaCryptoError> {
```

Comment:
I assume this is for deserializing a stored `BigInt`. Does this assume that the serialized `BigInt` was positive?

Reply:
Yes but like openssl it will handle the negative case too if the number starts with "-"